Remove some downcasting in Element

annevk opened the pull request:
#### cedfed3cd60b577ee81c25328e0fa1dd7cf53f3c
<pre>
Remove some downcasting in Element
<a href="https://bugs.webkit.org/show_bug.cgi?id=267585">https://bugs.webkit.org/show_bug.cgi?id=267585</a>

Reviewed by Chris Dumez.

Instead of shouldObtainBoundsFromSVGModel() that returns a boolean,
have elementWithSVGLayoutBox() that returns a RefPtr. Deduplicates
downcasting and clarifies the code.

* Source/WebCore/dom/Element.cpp:
(WebCore::elementWithSVGLayoutBox):
(WebCore::Element::boundsInRootViewSpace):
(WebCore::Element::absoluteEventBounds):
(WebCore::Element::getClientRects):
(WebCore::Element::boundingAbsoluteRectWithoutLayout const):
(WebCore::shouldObtainBoundsFromSVGModel): Deleted.

Canonical link: <a href="https://commits.webkit.org/273137@main">https://commits.webkit.org/273137@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/6b40811796596b9dab15121b9d45c5703140b95e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/34408 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/13220 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/36404 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/37090 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/31122 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/35491 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/15604 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/10332 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/30119 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/34923 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/11184 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/30681 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/9775 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/9856 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/30734 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/38373 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/31247 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/31050 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/35935 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/9985 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/7863 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/33882 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/11791 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/7909 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/10539 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/10833 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1637,13 +1637,13 @@ int Element::scrollHeight()
     ASSERT(element);
-    if (RefPtr svg = dynamicDowncast<SVGElement>(element))
-        return svg->hasAssociatedSVGLayoutBox();
+    if (RefPtr svg = dynamicDowncast<const SVGElement>(element); svg && svg->hasAssociatedSVGLayoutBox())
```
I doubt we need the `const` here? dynamicDowncast<>() is supposed to match constness already.

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1637,13 +1637,13 @@ int Element::scrollHeight()
 }
 
-inline bool shouldObtainBoundsFromSVGModel(const Element* element)
+inline RefPtr<const SVGElement> elementWithSVGLayoutBox(const Element* element)
```
Should take in a C++ reference.

cdumez commented on Source/WebCore/dom/Element.cpp:
```diff
@@ -1637,13 +1637,13 @@ int Element::scrollHeight()
-inline bool shouldObtainBoundsFromSVGModel(const Element* element)
+inline RefPtr<const SVGElement> elementWithSVGLayoutBox(const Element* element)
 {
     ASSERT(element);
```
Should be dropped.

cdumez reviewed (approved):
Nice.

annevk committed:
Remove some downcasting in Element
https://bugs.webkit.org/show_bug.cgi?id=267585

Reviewed by Chris Dumez.

Instead of shouldObtainBoundsFromSVGModel() that returns a boolean,
have elementWithSVGLayoutBox() that returns a RefPtr. Deduplicates
downcasting and clarifies the code.

* Source/WebCore/dom/Element.cpp:
(WebCore::elementWithSVGLayoutBox):
(WebCore::Element::boundsInRootViewSpace):
(WebCore::Element::absoluteEventBounds):
(WebCore::Element::getClientRects):
(WebCore::Element::boundingAbsoluteRectWithoutLayout const):
(WebCore::shouldObtainBoundsFromSVGModel): Deleted.

Canonical link: https://commits.webkit.org/273137@main
